### DISCUSSION TITLE
How to add a custom contract Serializer (DataSet .NET 4.8 is not compatible with DataSet .NET 7.0)

### DISCUSSION THREAD

**Discussion by alexvazquez** [CoreWCF/CoreWCF#1078] [General]
I was debuging all day trying to find out why my CoreWCF Service in .NET 7 was not working with my Client in .NET 4.8

My operation contract is:

```
[OperationContract]
DataSet GetMemberInfo(string uid, bool onDemand = false);
```

Everything seems to be working fine until the last step in my client where I got this error:

`InnerException: ArgumentException "Column requires a valid DataType".`

So that means that the DataTables inside my DataSet had some invalid DataType, but that was very strange because I didn't change any logic in the CoreWCF server side to get the DataSet from the Database.

After looking for 5 hours I found the problem. The problem is that when the DataSet is being serialized into XML to be sent, there are some DataTypes like GUID's where you get this:

`<xs:attribute name="addressuid" msdata:DataType="System.Guid, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e" type="xs:string"/>`

And if I compare it on how .NET 4.8 generates it:

`<xs:attribute name="addressuid" msdata:DataType="System.Guid, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" type="xs:string"/>`

So this totally makes sense, when .NET 4.8 tries to deserialize it into a DataSet it fails because it can't convert: **System.Guid, System.Private.CoreLib, Version=7.0.0.0** to  **System.Guid, mscorlib, Version=4.0.0.0**

I don't know if there is a correct way to workaround this, but maybe with a DataContract custom serialization where the string is replaced might work. I just did it manually and it worked perfectly but I need to have something implemented in my code to not have these issues.

Any help is well appreciated.

Thanks

**Comment by mconnew** [comment]
I suggest opening an issue in the dotnet/runtime repo as DataTable and DataSet classes both implement `IXmlSerializable`. This means those classes are explicitly choosing to take full control of xml serialization. The implementation has chosen how to serialize primitives such as Guid's (primitives from the perspective of XmlSerializer) instead of deferring that to XmlSerializer. In doing so they broke backwards compatibility with .NET Framework by explicitly encoding the .NET data types in the XML. The XmlSerializer goes out of its way to not do this for this exact reason.  

From a pragmatic standpoint, the easiest way (and it's still not a small task) would be to create your own type which wraps DataSet, and another one which wraps DataTable. Your types will also implement IXmlSerializable. When IXmlSerializable.WriteXml is called on your class, call the wrapped classes WriteXml and modify it. There are 2 ways to do this, efficiently, or with minimal code.  

The efficient way would be to create your own type derived from XmlWriter which overrides every method and wraps the passed in XmlWriter. You can't just override abstract methods, you need to override everything virtual too. Simply pass through all calls except those to do with attributes where you modify any `msdata:DataType` attributes to have the correct data before calling the wrapped method. Call the wrapped DataSet/DataType WriteXml method with this wrapping XmlWriter.  

The minimal code method is to instantiate an XmlWriter which writes to a TextWriter and calls the wrapped DataSet/DataType WriteXml method. Then do a string replace on the TextWriter. Maybe using a compiled regex? You might want to check performance of those two options. Once you have the Xml fixed up in a string, create an XmlReader using a TextReader around the fixed string. Then use XmlWriter.WriteNode which takes an XmlReader.  

If you need to read the data, you need to do the whole thing again for ReadXml. If you are exposing a WSDL from your CoreWCF service, you will also need to supply an implementation of IXmlSerializable.GetSchema to fixup namespaces. I had a quick look at DataSet.WriteXml and it has an option to emit the schema as part of serialization, so you might need to fix something there too.

**Reply by alexvazquez** [reply]
> I suggest opening an issue in the dotnet/runtime repo as DataTable and DataSet classes both implement `IXmlSerializable`. This means those classes are explicitly choosing to take full control of xml serialization. The implementation has chosen how to serialize primitives such as Guid's (primitives from the perspective of XmlSerializer) instead of deferring that to XmlSerializer. In doing so they broke backwards compatibility with .NET Framework by explicitly encoding the .NET data types in the XML. The XmlSerializer goes out of its way to not do this for this exact reason.
> 
> From a pragmatic standpoint, the easiest way (and it's still not a small task) would be to create your own type which wraps DataSet, and another one which wraps DataTable. Your types will also implement IXmlSerializable. When IXmlSerializable.WriteXml is called on your class, call the wrapped classes WriteXml and modify it. There are 2 ways to do this, efficiently, or with minimal code.
> 
> The efficient way would be to create your own type derived from XmlWriter which overrides every method and wraps the passed in XmlWriter. You can't just override abstract methods, you need to override everything virtual too. Simply pass through all calls except those to do with attributes where you modify any `msdata:DataType` attributes to have the correct data before calling the wrapped method. Call the wrapped DataSet/DataType WriteXml method with this wrapping XmlWriter.
> 
> The minimal code method is to instantiate an XmlWriter which writes to a TextWriter and calls the wrapped DataSet/DataType WriteXml method. Then do a string replace on the TextWriter. Maybe using a compiled regex? You might want to check performance of those two options. Once you have the Xml fixed up in a string, create an XmlReader using a TextReader around the fixed string. Then use XmlWriter.WriteNode which takes an XmlReader.
> 
> If you need to read the data, you need to do the whole thing again for ReadXml. If you are exposing a WSDL from your CoreWCF service, you will also need to supply an implementation of IXmlSerializable.GetSchema to fixup namespaces. I had a quick look at DataSet.WriteXml and it has an option to emit the schema as part of serialization, so you might need to fix something there too.

Thanks @mconnew , read your comment and seems not an easy thing. For now what I did is to use my custom Gzip compressor behavior and before doing the compression I'm converting the stream to XmlDoc and then do the replacement there.  I tried with a XML size of about 5mb and took 43 milliseconds to do the replacement, so not bad.

I really thought there was like a way to create my custom data contract serializer and override some methods. For sure your recommendation sounds to be the cleanest and professional but will take time for me to understand all that for sure.

Anyway I appreciate so much your comment and if you think there other way please let me know.

**Comment by mconnew** [comment]
Going via XmlDocument is also a viable option. My biggest concern is the GC overhead. Allocating large continuous amounts of memory (eg in a String or MemoryStream) which isn't pooled will cause a hidden cost which you won't see until there's a little bit of memory pressure. Then you might get long intermittent GC pauses (like 1 second long).

**Reply by alexvazquez** [reply]
> Going via XmlDocument is also a viable option. My biggest concern is the GC overhead. Allocating large continuous amounts of memory (eg in a String or MemoryStream) which isn't pooled will cause a hidden cost which you won't see until there's a little bit of memory pressure. Then you might get long intermittent GC pauses (like 1 second long).

Actually I'm doing the XmlDocument thing in my GzipMessageEncoderFactory.

I have this:

`static ArraySegment<byte> CompressBuffer(ArraySegment<byte> buffer, BufferManager bufferManager, int messageOffset)
            {

                var stopwatch = Stopwatch.StartNew();

                // Replace the assembly reference in the buffer with the .NET Framework format
                string bufferString = Encoding.UTF8.GetString(buffer.Array, messageOffset, buffer.Count);

                XmlDocument xmlDoc = new XmlDocument();
                xmlDoc.LoadXml(bufferString);

                // IMPORTANT!!
                // Special logic to make compatible .NET 7 to .NET 4.8 GUID Serialization  msdata:DataType="System.Guid, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"

                // Process the XML and update Guid serialization
                XmlNamespaceManager nsMgr = new XmlNamespaceManager(xmlDoc.NameTable);
                nsMgr.AddNamespace("xs", "http://www.w3.org/2001/XMLSchema");
                nsMgr.AddNamespace("msdata", "urn:schemas-microsoft-com:xml-msdata");

                var guidAttributes = xmlDoc.SelectNodes("//xs:attribute[@msdata:DataType='System.Guid, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e']", nsMgr);

                foreach (XmlNode guidAttribute in guidAttributes)
                {
                    if (guidAttribute.Attributes?["msdata:DataType"] != null)
                    {
                        guidAttribute.Attributes["msdata:DataType"].Value = "System.Guid, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089";
                    }
                }

                byte[] compressedBuffer = CompressString(xmlDoc.InnerXml);

                stopwatch.Stop();

                // Return the compressed buffer as an ArraySegment<byte>
                int compressedSize = compressedBuffer.Length;
                byte[] compressedMessage = bufferManager.TakeBuffer(compressedSize + messageOffset);
                Array.Copy(buffer.Array, 0, compressedMessage, 0, messageOffset);
                Array.Copy(compressedBuffer, 0, compressedMessage, messageOffset, compressedSize);
                return new ArraySegment<byte>(compressedMessage, 0, messageOffset + compressedSize);
            }`

And yes Im using MemoryStream but inside a using so I think that will dispose the object and GC no problem right?
```

  static byte[] CompressString(string input)
            {
                byte[] inputBytes = Encoding.UTF8.GetBytes(input);
                using (MemoryStream memoryStream = new MemoryStream())
                {
                    using (GZipStream gzipStream = new GZipStream(memoryStream, CompressionMode.Compress))
                    {
                        gzipStream.Write(inputBytes, 0, inputBytes.Length);
                    }
                    return memoryStream.ToArray();
                }
            }


```

**Reply by mconnew** [reply]
MemoryStream isn't great. It starts off with a buffer the size of the first write. Then when it's full, it allocates a new array with a size double the previous one, then rounded up to the next power of 2, then copies the data across. When you are doing 5MB, it's likely allocated an 8MB array for that. And it's discarded arrays which are 4MB, 2MB, 1MB, 512K, 256K, and 128K. The ones smaller than that (64K and smaller) are all allocated on the heap. Anything bigger than 85000 bytes gets allocated in the Large Object Heap which only gets collected when you do a gen2 collection, and the LOH doesn't get compacted and the allocation algorithm is search for the first gap big enough. You run the risk of having a large LOH with a lot of fragmentation. Then the final call to `MemoryStream.ToArray()` allocates yet another array which is the correct size for its content (instead of 8MB) and copies the data over. You're better off calling TryGetBuffer which returns the underlying array and return an ArraySegment to specify how big the contents are.  
I recommend copying the implementation of [BufferManagerOutputStream](https://github.com/CoreWCF/CoreWCF/blob/main/src/CoreWCF.WebHttp/src/CoreWCF/Channels/BufferManagerOutputStream.cs) and [BufferedOutputStream](https://github.com/CoreWCF/CoreWCF/blob/df8e34d29201091b3a2ad491b44f1167be40786b/src/CoreWCF.Primitives/src/CoreWCF/Runtime/BufferedOutputStream.cs) from the CoreWCF source code (just maintain the copyright header and license is MIT so permissive with no release of your code required). It's BufferManager aware and uses it to allocate new buffers in chunk with no copy on resize. Where's an annoying abstraction with an InternalBufferManager which I don't know why it is there, but you can change the code to use BufferManager directly and strip out that code. The array returned from BufferedOutputStream.ToArray returns an array pulled from the BufferManager so can be returned. This should fix any potential GC problems.

**Reply by alexvazquez** [reply]
Wow very interesting, sometimes as C# developer you use methods that you look everywhere but you never think about efficiency and memory optimization.  Of course simplicity of methods have always a trade-off. I will take a look on those classes and see how its implemented. 

Appreciate so much!
